### PR TITLE
Enforce safe socket paths, reject TCP addresses (issue #538)

### PR DESCRIPTION
**Pull request by ojongerius**
## What

Add socket-path safety validation to the daemon startup. The daemon now:

1. **Rejects unsafe socket paths by default** — `--socket` / `AGENTRECEIPTS_SOCKET` must resolve under per-platform safe directories (Linux: `$XDG_RUNTIME_DIR`, `/run`, `/var/run`; macOS: `$TMPDIR`, `/var/run`, `$XDG_DATA_HOME/agent-receipts`), or the daemon refuses to start.

2. **Rejects TCP addresses unconditionally** — even with `--unsafe-socket-path`, TCP loopback addresses are forbidden per ADR-0010 § IPC transport (the daemon speaks Unix-domain sockets only).

3. **Permits deliberate overrides with `--unsafe-socket-path`** — when passed, unsafe paths are allowed but the daemon warns on startup and every 60 seconds for the lifetime of the process.

4. **Canonicalizes paths before checking** — symlinks are resolved with `filepath.EvalSymlinks`, so a symlink pointing out of the safe set is judged by its real target, not its apparent location. Fails closed: unresolvable paths (permission errors, symlink loops) are treated as unsafe.

## Why

Issue #538 identified that a socket in a shared, world-traversable, periodically-swept directory (e.g. `/tmp`) defeats the per-user trust boundary that peer-credential capture depends on. The safe default socket path always resolves under a per-user runtime/data directory, but an explicit override could silently abandon that safety. This change prevents that by enforcing the safe-socket-location policy at startup, with a documented escape hatch for legitimate use cases (containers with unusual mounts, dev experiments, integration tests that need short paths to stay within the 104-byte `AF_UNIX sun_path` limit on macOS).

## Implementation

- **`socketpolicy.go`** (new): Core validation logic
  - `checkSocketPath()`: Main entry point; distinguishes safe paths, unsafe paths with flag, and rejected paths (TCP or unsafe without flag)
  - `allowedSocketRoots()`: Per-platform safe directory list
  - `isSocketPathSafe()`: Canonicalization and containment check
  - `canonicalizePath()`: Resolves symlinks, handling non-existent socket files
  - `pathWithin()`: Checks if a path is under a root
  - `looksLikeTCPAddress()`: Detects TCP addresses (bare `host:port`, `tcp://` scheme)
  - `warnUnsafeSocketPath()`: Emits startup and periodic warnings

- **`socketpolicy_test.go`** (new): Comprehensive test coverage
  - Default paths are safe
  - Explicit safe paths are allowed
  - Unsafe paths are rejected without flag, allowed with flag
  - TCP addresses rejected unconditionally
  - Symlink edge cases (escape, loops) fail closed
  - Warning emission (startup-only vs. periodic)

- **`daemon.go`**: Integrate validation into `Run()`
  - Add `UnsafeSocketPath` config field
  - Call `checkSocketPath()` before binding the socket
  - Launch warning goroutine if unsafe path is permitted

- **`cmd/agent-receipts-daemon/main.go`**: Add `--unsafe-socket-path` flag

- **Test fixtures** (`tests_framework.go`, `integration_test.go`, `emitter_test.go`): Set `UnsafeSocketPath: true` since they use short socket paths under `/tmp` to stay within the `sun_path` limit

- **Documentation** (`README.md`, `CHANGELOG.md`): Explain the policy, safe directories, and escape hatch

## Checklist

- [x] Tests pass for all changed components
- [x] Linter passes (`go vet`)
- [x] No real keys or secrets in the diff
- [x] Cross-language tests pass (integration tests updated to opt into escape hatch)

## Security

- [x] This PR is itself a security hardening (closes the "unsafe configuration silently accepted" gap from issue #538). It adds no new attack surface: the only new input is the `--unsafe-socket-path` flag, validation fails closed on unresolvable paths, and TCP addresses are rejected unconditionally regardless of the flag. A self-review confirmed there is no path-canonicalization bypass (`..` traversal and symlink escapes are resolved before the containment check) and that the default socket path always remains inside the safe set.

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -286,6 +292,19 @@ func Run(ctx context.Context, cfg Config) error {
...
+	// an unsafe override without --unsafe-socket-path refuses to start here;
+	// an unsafe override with the flag starts but warns on a 60s cadence for
+	// the daemon's lifetime.
+	unsafeSocket, err := checkSocketPath(cfg.SocketPath, cfg.UnsafeSocketPath)
+	if err != nil {
+		return err
+	}
+	if unsafeSocket {
+		go warnUnsafeSocketPath(ctx, cfg.Logger, cfg.SocketPath, unsafeSocketWarnInterval)
+	}
```
[the remark this answers is not in the document]
Comment (ojongerius) [reply daemon/daemon.go:312]: Good catch — fixed in e02f0b8. The warning goroutine now runs on a `context.WithCancel(ctx)` with a `defer cancelWarn()`, so it's always cancelled when `Run` returns (including a later startup error), and can no longer outlive a failed start.

File: daemon/socketpolicy_test.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.buf.Reset()
+}
+
+// setupSocketEnv pins every environment variable that allowedSocketRoots and
+// DefaultSocketPath consult to test-controlled temp dirs, none of which is
+// /tmp, so the safe set is deterministic across hosts. It also clears
+// AGENTRECEIPTS_SOCKET so DefaultSocketPath resolves the platform default.
+// Returns the per-platform directory a safe explicit override may live under.
```
[the remark this answers is not in the document]
Comment (ojongerius) [reply daemon/socketpolicy_test.go]: Agreed — the comment was inaccurate. Reworded in e02f0b8: the invariant is that the safe roots are distinct, test-controlled temp subdirectories (not host-dependent), and the sibling `/tmp` unsafe path is never inside one. Fixed the matching comment in `unsafeRoot` too.

File: daemon/socketpolicy.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	if err != nil {
+		return false
+	}
+	if _, err := strconv.Atoi(port); err != nil {
+		return false
+	}
+	if host == "" || strings.EqualFold(host, "localhost") {
+		return true
+	}
+	return net.ParseIP(host) != nil
```
[the remark this answers is not in the document]
Comment (ojongerius) [reply daemon/socketpolicy.go]: Fixed in b49148c. `looksLikeTCPAddress` now treats any bare `host:port` with a numeric port and no path separator as TCP, regardless of host — so `example.com:9000` (and any other hostname) is rejected unconditionally, matching the contract. Real daemon sockets are absolute paths (they contain a separator) and short-circuit before the `host:port` branch, so no legitimate config is misflagged. Added test coverage for non-IP hostnames.

File: daemon/socketpolicy.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+		return true
+	}
+	return rel != ".." && !strings.HasPrefix(rel, ".."+string(filepath.Separator))
+}
+
+// looksLikeTCPAddress reports whether s is a TCP socket address rather than a
+// Unix-domain path. It matches an explicit tcp/tcp4/tcp6 URL scheme, or a bare
+// host:port with a numeric port whose host is empty, "localhost", or an IP
+// literal (covers ":9000", "127.0.0.1:9000", "[::1]:9000", "localhost:9000").
+// Unix paths contain a path separator and never reach the host:port branch, so
```
[the remark this answers is not in the document]
Comment (ojongerius) [reply daemon/socketpolicy.go]: Updated in b49148c — the doc comment now states that any `host:port` form is matched regardless of host (no longer restricted to empty/localhost/IP), and explains why absolute Unix paths can't be misflagged.

File: daemon/socketpolicy.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+// any bare host:port with a numeric port and no path separator — regardless of
+// host (covers ":9000", "127.0.0.1:9000", "[::1]:9000", "localhost:9000",
+// "example.com:9000"). The host is deliberately not restricted to loopback: the
+// daemon never speaks TCP, so every host:port form is rejected uniformly rather
+// than only loopback variants, keeping checkSocketPath's "TCP rejected
+// unconditionally" contract honest.
+//
+// A Unix path is anchored by a path separator and short-circuits before the
+// host:port branch, and a real daemon socket is always an absolute path, so
+// this cannot misflag a legitimate configuration.
```
[the remark this answers is not in the document]
Comment (ojongerius) [reply daemon/socketpolicy.go]: Adjusted the comment in c47baf1 rather than adding enforcement (keeping this PR's scope minimal). It no longer claims absoluteness is an invariant — it now notes that `validateConfig` doesn't enforce it, and that the only Unix path reaching the `host:port` branch is a bare relative filename shaped like `host:port` with a numeric port (e.g. `foo:9000`), which isn't a real socket configuration, so treating it as TCP is acceptable rather than a misflag.

I considered enforcing absolute socket paths in `validateConfig`, but that's a broader behavior change (and `DBPath`/`KeyPath` aren't enforced either), so I'd rather keep it out of this PR. Happy to file a follow-up if you'd like absolute-path enforcement across the daemon's path config.
